### ISSUE TITLE
[$250] mWeb/Safari - WS Switcher - The back button does not respond when switching between workspaces

### ISSUE THREAD

**Issue by IuliiaHerets** [Expensify/App#54527]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** v9.0.78-1
**Reproducible in staging?:** Yes
**Reproducible in production?:** Yes
**Issue was found when executing this PR:** https://github.com/Expensify/App/pull/54030
**Issue reported by:** Applause Internal Team
**Device used:** iPhone 13
**App Component:** Other

## Action Performed:
1.Open the app
2.Go to the WS switcher
3.Switch to/Filter by a workspace
4.Use device back button/gesture

## Expected Result:
The app doesn't close but the app navigates to the previous workspace

## Actual Result:
The back button does not respond when switching between workspaces

## Workaround:
Unknown

## Platforms:
- [ ] Android: Standalone
- [ ] Android: HybridApp
- [ ] Android: mWeb Chrome
- [x] iOS: Standalone
- [ ] iOS: HybridApp
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/3b8f1bd7-deb7-47ff-b70a-9769d01fc671

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021876256758509266830</li>
        <li>Upwork Job ID: 1876256758509266830</li>
        <li>Last Price Increase: 2025-02-17</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @s77rt</details>

**Comment by sonialiap** [comment]
Working part time through the holidays and prioritizing customer issues. Will look at this ASAP

**Comment by s77rt** [comment]
Still waiting for proposals

**Comment by pac-guerreiro** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

1. User goes to workspace switcher page, then the history stack will be like this `home > workspace switcher`
2. User then selects a workspace and navigates back, then the history stack will be like this `home` instead of `home > workspace switcher > home`.

### What is the root cause of that problem?

After selecting a workspace, the user navigates back to the previous page and the workspace switcher page is removed from history.

### What changes do you think we should make in order to solve the problem?


Instead of navigating the user back, we can push the previous screen into the history stack so that the user is able to go back to workspace switcher page using browser's go back button.

### What specific scenarios should we cover in automated tests to prevent reintroducing this issue in the future?


### What alternative solutions did you explore? (Optional)

**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

**Comment by s77rt** [comment]
@pac-guerreiro Thanks for the proposal. I don't think your proposal aligns with the expected results. On going back we should still be on the home page but with the selected workspace being the previous one.

**Comment by pac-guerreiro** [comment]
@s77rt If that's the case, then I think the page url would need to include the current workspace id so that we could load the previous state properly

**Comment by s77rt** [comment]
@pac-guerreiro This appears to be already working as expected on web. Can you check why this is not the case for mWeb-Safari?

**Comment by pac-guerreiro** [comment]
@s77rt you're right, what I proposed in my previous comment is already implemented 😅 I need to do some testing to check what's happening

**Comment by pac-guerreiro** [comment]
@s77rt I am also able to reproduce this on iOS chrome

https://github.com/user-attachments/assets/bfaab2a1-566a-4dda-9d1d-516c428c6fbd

**Comment by pac-guerreiro** [comment]
@s77rt here is something interesting that I found while testing the web on a desktop:

https://github.com/user-attachments/assets/cc420362-5596-4a27-8ec5-a9b86c84575d

It seems that we're applying different logic to how we navigate for mobile specifically. I need to dig through the code to confirm what's different.

**Comment by pac-guerreiro** [comment]
I think the problem lies here:

<img width="1303" alt="Image" src="https://github.com/user-attachments/assets/44d210b6-898f-45ee-9ab1-90becc60e41c" />

On larger screens, the central pane is focused when we change workspaces but on smaller screens it's not.

So we end up having this scenarios happening:

**Smaller Screen**

`/home` -> `/workspace-switcher` -> `/home`

**Larger Screen**

`/r/12345` -> `/workspace-switcher` -> `/w/1/r/12345`

When the first link is the same as the last one, then we can't go back, otherwise we can.

Tomorrow I'll try to find a solution for the problem!

**Comment by pac-guerreiro** [comment]
Btw I’m Pedro Guerreiro from Callstack - expert contributor group.

**Comment by pac-guerreiro** [comment]
@s77rt I was wrong about the culprit, it seems we handle navigation logic to switch workspace here:

<img width="1218" alt="Image" src="https://github.com/user-attachments/assets/c3308f6e-b454-4d08-b809-3c287b8c9d00" />

**Comment by pac-guerreiro** [comment]
I also noticed that we get `activeWorkspaceID` in two different ways: 

- `useActiveWorkspaceFromNavigationState`: as the name explains, this returns the active workspace ID from navigation params
- `useActiveWorkspace`: this returns the active workspace ID from a ContextProvider

None of these hooks seems to read the `policyID` correctly from url with `/w/:policyID/home` pattern.

IMO we should read the `activeWorkspaceID` from route instead of storing it in a context or relying on params passed to navigation, at least on web.

Nonetheless, I'll continue investigating possible solutions to solve this

**Comment by s77rt** [comment]
Not overdue

**Comment by pac-guerreiro** [comment]
@s77rt 

These are today's findings:

The browser history is synched with the app stack navigator routes. When open the app with the narrow layout there's only one route on the stack navigator, which is the bottom tab navigator. The bottom tab navigator contains it's own routes, which in this situation will consist in the `/home` route. When we switch workspace, we push a new route into bottom tab navigator so that it displays the updated home screen. But the stack navigator still has a single route, which is the bottom tab navigator, so we're not able to go back in the browser. In the wide layout, the stack navigator contains two routes: bottom tab navigator and report. When we switch workspace, we push a new report route into stack navigator which enables us to go back in the browser history.

In theory, what I need to do is to push something into the stack navigator routes to enable going back in the browser. I'm just not sure what yet, because pushing a new bottom tab navigator might break the app.

Tomorrow I'll do more research and try putting my theory into practise.

**Comment by pac-guerreiro** [comment]
Today I tried different approaches to solve the issue but ultimately the only one that seems to work is to push a new bottom tab navigator into the stack navigator, which from my testing seems to affect app performance. At this stage I don't think we can fix this issue without hacks that will affect UX.

**Comment by pac-guerreiro** [comment]
This issue was also present in version `9.0.0-0`

**Comment by s77rt** [comment]
Yeah, this is not a new bug. But let's keep this open for now

**Comment by s77rt** [comment]
Same ^

**Comment by s77rt** [comment]
Same ^

**Comment by s77rt** [comment]
Same ^

**Comment by s77rt** [comment]
Still looking for proposals

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by s77rt** [comment]
Still looking for proposals

**Comment by pac-guerreiro** [comment]
@s77rt this [PR](https://github.com/Expensify/App/pull/49539) got merged yesterday and it might fix this issue. I'll investigate it!

**Comment by s77rt** [comment]
@pac-guerreiro Thanks you! Looking forward for your findings

**Comment by s77rt** [comment]
@pac-guerreiro Thanks for checking. Given that the original issue is no longer reproducible I think we can close this one and the other issue can be handled as follow ups for the navigation refactor

**Comment by s77rt** [comment]
@sonialiap Let's close this one

**Comment by s77rt** [comment]
@sonialiap We can close this

**Comment by pac-guerreiro** [comment]
@s77rt regarding the follow up issues, can I be tagged on them?

**Comment by s77rt** [comment]
cc @mountiny Do we have an issue for fixing surfaced issues from https://github.com/Expensify/App/pull/49539

**Comment by mountiny** [comment]
@s77rt sorry, can you explain what you mean? We do track follow up issues by adding the `[Navigation]` prefix and SWM is going through them

**Comment by s77rt** [comment]
@pac-guerreiro Can you still reproduce this bug https://github.com/Expensify/App/issues/54527#issuecomment-2654130504?

Just tested and seems to be working well:

https://github.com/user-attachments/assets/43a6f480-fcad-466c-af5a-d80d26f62518

**Comment by pac-guerreiro** [comment]
https://github.com/user-attachments/assets/93de04f9-5378-4dbe-85c6-5cb3a51b3728

@s77rt the issue where the home screen is shown as blank for a few seconds seems to be fixed, but moving in the history after will not show the correct workspace just home instead

**Comment by s77rt** [comment]
@pac-guerreiro Got it, thanks! I have reported this here https://expensify.slack.com/archives/C049HHMV9SM/p1739971259098259

**Comment by s77rt** [comment]
https://github.com/Expensify/App/issues/57121

**Comment by s77rt** [comment]
@sonialiap We can close this

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Second week)

**Comment by s77rt** [comment]
This can be closed

**Comment by pac-guerreiro** [comment]
@mountiny this issue can be closed 😄
